Realm/RLMObjectStore.mm

```diff
@@ -95,45 +95,178 @@ void RLMInitializeSwiftAccessorGenerics(__unsafe_unretained RLMObjectBase *const
+    }
+}
+
+/* If a row exists with the specified primary key value, return its index. Otherwise, return `realm::not_found`.
+ *
+ * Precondition: `info` must refer to a class which has a primary key property
+ * Precondition: `primaryValue` is a validated property value that has been coerced to `nil`
+ */
+static NSUInteger getRowForObjectWithPrimaryKey(RLMClassInfo const& info, id primaryValue) {
+    REALM_ASSERT_DEBUG(info.propertyForPrimaryKey());
+
+    RLMProperty *const primaryProperty = info.propertyForPrimaryKey();
+    const NSUInteger primaryPropertyColumn = info.tableColumn(primaryProperty);
+
+    try {
+        switch (primaryProperty.type) {
+            case RLMPropertyTypeString:
+                return info.table()->find_first_string(primaryPropertyColumn, RLMStringDataWithNSString(primaryValue));
+
+            case RLMPropertyTypeInt:
+                if (primaryValue) {
+                    return info.table()->find_first_int(primaryPropertyColumn, [primaryValue longLongValue]);
+                } else {
+                    return info.table()->find_first_null(primaryPropertyColumn);
+                }
+
+            default:
+                REALM_UNREACHABLE();
         }
-        created = YES;
+
+    }
+    catch (std::exception const& e) {
+        @throw RLMException(e);
     }
+}
 
-    // get accessor
+/* Create a row with the specified primary key value and return its index.
+ *
+ * Precondition: `info` must refer to a class which has a valid primary key property
+ * Precondition: a write transaction is in progress
+ * Precondition: no row already exists with the specified `primaryValue` for this model
+ */
+static NSUInteger createRowForObjectWithPrimaryKey(RLMClassInfo const& info, id primaryValue) {
+    REALM_ASSERT_DEBUG(info.propertyForPrimaryKey());
+    REALM_ASSERT_DEBUG(info.realm.inWriteTransaction);
+    REALM_ASSERT_DEBUG(getRowForObjectWithPrimaryKey(info, primaryValue) == realm::not_found);
+
+    RLMProperty *const primaryProperty = info.propertyForPrimaryKey();
+    const NSUInteger primaryColumnIndex = info.tableColumn(primaryProperty);
+
+    // create row
+    const NSUInteger rowIndex = createRowForObject(info);
+    Row row = info.table()->get(rowIndex);
+
+    // set value for primary key
+    validateValueForProperty(primaryValue, primaryProperty);
+    primaryValue = RLMCoerceToNil(primaryValue);
+
+    try {
+        switch (primaryProperty.type) {
+            case RLMPropertyTypeString:
+                REALM_ASSERT_DEBUG(!primaryValue || [primaryValue isKindOfClass:NSString.class]);
+                row.set_string_unique(primaryColumnIndex, RLMStringDataWithNSString((NSString *)primaryValue));
+                break;
+
+            case RLMPropertyTypeInt:
+                if (primaryValue) {
+                    REALM_ASSERT_DEBUG([primaryValue isKindOfClass:NSNumber.class]);
+                    row.set_int_unique(primaryColumnIndex, ((NSNumber *)primaryValue).longLongValue);
```

Comment:
I guess. It seems safer to me to *always* cast away from `id`, otherwise you don't have compile-time verification that the object can respond to the message you are sending. I could've easily called a similarly named method that's defined on another type, not `NSNumber`.